lib/jstp.js

```diff
@@ -0,0 +1,108 @@
+'use strict';
+
+const jstp = require('@metarhia/jstp');
+const { codes: errorCodes } = require('./errors');
+
+const validate = {
+  number: value => typeof value === 'number',
+  string: value => typeof value === 'string',
+  object: value => typeof value === 'object' && !Array.isArray(value),
+  objectArray: value => Array.isArray(value) && value.every(validate.object),
+};
+
+const createApi = gsProvider => {
+  let nextCursorId = 0;
+  const api = {
+    provider: {
+      get(connection, id, callback) {
+        if (!validate.string(id)) {
+          callback(jstp.ERR_INVALID_SIGNATURE);
+          return;
+        }
+        gsProvider.get(id, (err, record) => {
+          callback(err && err.code, record);
+        });
+      },
+
+      set(connection, record, callback) {
+        if (!validate.object(record)) {
+          callback(jstp.ERR_INVALID_SIGNATURE);
+          return;
+        }
+        gsProvider.set(record, err => {
+          callback(err && err.code);
+        });
+      },
+
+      create(connection, category, record, callback) {
+        if (!validate.string(category) || !validate.object(record)) {
+          callback(jstp.ERR_INVALID_SIGNATURE);
+          return;
+        }
+        gsProvider.create(category, record, (err, id) => {
+          callback(err && err.code, id && id.toString());
+        });
+      },
+
+      update(connection, category, query, patch, callback) {
+        if (
+          !validate.string(category) ||
+          !validate.object(query) ||
+          !validate.object(patch)
+        ) {
+          callback(jstp.ERR_INVALID_SIGNATURE);
+          return;
+        }
+        gsProvider.update(category, query, patch, (err, count) => {
+          callback(err && err.code, count);
+        });
+      },
+
+      delete(connection, category, query, callback) {
+        if (!validate.string(category) || !validate.object(query)) {
+          callback(jstp.ERR_INVALID_SIGNATURE);
+          return;
+        }
+        gsProvider.delete(category, query, (err, count) => {
+          callback(err && err.code, count);
+        });
+      },
+
+      select(connection, category, query, callback) {
+        if (!validate.string(category) || !validate.object(query)) {
+          callback(jstp.ERR_INVALID_SIGNATURE);
+          return;
+        }
+        const cursorId = nextCursorId++;
+        connection.session.set(
+          `cursor${cursorId}`,
+          gsProvider.select(category, query)
+        );
+        callback(null, cursorId);
+      },
+    },
+
+    cursor: {
+      fetch(connection, cursorId, jsql, callback) {
+        if (!validate.number(cursorId) || !validate.objectArray(jsql)) {
```

Comment:
On a second thought, could we perhaps change the api to allow `jsql` to be null to avoid creating an empty array every time?